src/main/java/com/redhat/devtools/lsp4ij/features/codeBlockProvider/LSPCodeBlockProvider.java

```diff
@@ -0,0 +1,201 @@
+import org.jetbrains.annotations.Nullable;
+
+import java.util.*;
+
+/**
+ * Code block provider that uses information from {@link LSPSelectionRangeSupport} and {@link LSPFoldingRangeBuilder}.
+ */
+public class LSPCodeBlockProvider implements CodeBlockProvider {
+
+    @Override
+    @Nullable
+    public TextRange getCodeBlockRange(Editor editor, PsiFile file) {
+        Document document = editor.getDocument();
+        CharSequence documentChars = document.getCharsSequence();
+        int documentLength = documentChars.length();
+
+        // Adjust the offset slightly based on before/after brace to ensure evaluation occurs "within" the braced block
+        int offset = editor.getCaretModel().getOffset();
+        Character beforeCharacter = offset > 0 ? documentChars.charAt(offset - 1) : null;
+        Character afterCharacter = offset < documentLength ? documentChars.charAt(offset) : null;
+        if (LSPCodeBlockUtils.isCodeBlockStartChar(file, afterCharacter)) {
+            offset++;
+        } else if (LSPCodeBlockUtils.isCodeBlockEndChar(file, beforeCharacter)) {
+            offset--;
+        }
+
+        // See if we're anchored by a known brace character
+        int openBraceOffset = -1;
+        Character openBraceChar = null;
+        int closeBraceOffset = -1;
+        Character closeBraceChar = null;
+        if ((offset > 0) && LSPCodeBlockUtils.isCodeBlockStartChar(file, documentChars.charAt(offset - 1))) {
+            openBraceOffset = offset - 1;
+            openBraceChar = documentChars.charAt(offset - 1);
+            closeBraceChar = LSPCodeBlockUtils.getCodeBlockEndChar(file, openBraceChar);
+        } else if (LSPCodeBlockUtils.isCodeBlockEndChar(file, documentChars.charAt(offset))) {
+            closeBraceOffset = offset;
+            closeBraceChar = documentChars.charAt(offset);
+            openBraceChar = LSPCodeBlockUtils.getCodeBlockStartChar(file, closeBraceChar);
+        } else if ((offset < (documentLength - 1)) && LSPCodeBlockUtils.isCodeBlockEndChar(file, documentChars.charAt(offset + 1))) {
+            closeBraceOffset = offset + 1;
+            closeBraceChar = documentChars.charAt(offset + 1);
+            openBraceChar = LSPCodeBlockUtils.getCodeBlockStartChar(file, closeBraceChar);
+        }
+
+        // Try to find it first using the selection ranges which tend to be more accurate; we must use the effective
+        // offset for selection ranges to act as if we're in the adjusted braced block
+        TextRange codeBlockRange = getUsingSelectionRanges(
+                file,
+                editor,
+                offset,
+                openBraceChar,
+                openBraceOffset,
+                closeBraceChar,
+                closeBraceOffset
+        );
+        if (codeBlockRange != null) {
+            return codeBlockRange;
+        }
+
+        // Failing that, try to find it using the folding ranges
+        return getUsingFoldingRanges(
```

Comment:
If we couldn't find a good code block from the selection ranges, try to find one using the folding ranges.